### PR TITLE
Give searches unique ID, use Mutex to guarantee 1 at a time

### PR DESCRIPTION
**Pull request by lorton**
We were seeing the wrong results returned sometimes, I *think* this is because the http server is run multithreaded and it could end up doing multiple search requests at once with a single linear pipe to communicate to backend.  The Mutex in python should prevent this from happening, and I've also added a unique ID identifier for each search that should serve as a fallback to make sure incorrect results are never returned.

### CODE REVIEW

File: python/gpusim_server.py [hunk 1]
```diff
@@ -95,17 +119,20 @@ def do_POST(self):
...
-            for i in range(return_count):
-                scores.append(data_reader.readFloat())
-            return return_count, smiles, ids, scores
+    def deserialize_results(self, request_num, output_qba):
+        data_reader = QtCore.QDataStream(output_qba)
+        returned_request = data_reader.readInt()
+        if request_num != returned_request:
+            raise RuntimeError("Incorrect result ID returned!")
+        return_count = data_reader.readInt()
+        smiles, ids, scores = [], [], []
```
Comment (cdvonbargen) [inline python/gpusim_server.py]: eh, you can do this all in one loop. list comprehensions also don't seem inappropriate.

File: python/gpusim_server.py [hunk 2]
```diff
@@ -73,16 +80,33 @@ def do_GET(self):
...
+                    request_num, output_qba)
+            except RuntimeError:
+                self.flush_socket(dbname)
+                raise
+
+            return smiles, ids, scores, src_smiles
+        finally:
+            search_mutex.unlock()
+
+    def flush_socket(self, dbname):
```
Comment (cdvonbargen) [inline python/gpusim_server.py]: what is dbname being used for?

File: python/gpusim_server.py [hunk 3]
```diff
@@ -73,16 +80,33 @@ def do_GET(self):
...
+        search_mutex.lock()
+        try:
+            if dbname not in sockets:
+                self.send_error(404, f'DB {dbname} not found in options: {sockets.keys()}') #noqa
+                raise ValueError('DB not found')
+            allsmiles, allids, allscores = [], [], []
+            src_smiles, return_count, similarity_cutoff = self.get_posted_data()
+            request_num = random.randint(0, 2**31)
+            print("Processing request {}".format(request_num), file=sys.stderr)
+            output_qba = self.get_data(dbname, src_smiles, return_count,
```
Comment (cdvonbargen) [inline python/gpusim_server.py]: It looks like the mutex try/finally only needs to be around get_data, no?

Comment (lorton) [reply python/gpusim_server.py]: Not really, coudl also be the dbname sockets check, get_ata or deserialize_results.  Really I want to make sure if anything happens we unlock the mutex, it's not blindly catching anything just always making sure the mutex is unlocked, so I think it's fine to have it at this higher level

File: python/gpusim_server.py (hunk 2 above)
Comment (lorton) [reply python/gpusim_server.py]: Well...this shows we haven't hit this code yet, as theoretically it should be being used where socket_name is...

File: python/gpusim_server.py (hunk 1 above)
Comment (lorton) [reply python/gpusim_server.py]: How can I do it all in 1 loop?  The pipe exports in order:
111112222233333
reading all in 1 loop with 1 read after another would be trying to read:
123123123123123

Which wouldn't match how the backend exports.

File: python/gpusim_server.py (hunk 3 above)
Comment (cdvonbargen) [reply python/gpusim_server.py]: Can you turn the mutex locking/unlocking into a context manager then?

File: python/gpusim_server.py (hunk 1 above)
Comment (cdvonbargen) [reply python/gpusim_server.py]: oh, duh, I see how it's a pipe. shame you didn't make a serializable results object :/

File: python/gpusim_server.py (hunk 2 above)
Comment (lorton) [reply python/gpusim_server.py]: Fixed

**Pull request merged**